pkg/plugins/builtin/ssh/ssh.go

```diff
@@ -226,19 +244,55 @@ trap printResultJSON EXIT
 		extraCmd = execStr
 	}
 
-	stdoutstderr, err := session.CombinedOutput(extraCmd)
-	if err != nil {
-		exitErr, ok := err.(*ssh.ExitError)
+	ch := make(chan sshReturn)
+	funcContext, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	// executing the SSH connection
+	go func(ch chan<- sshReturn) {
+		stdoutstderr, err := session.CombinedOutput(extraCmd)
+		select {
+		case <-funcContext.Done():
+		case ch <- sshReturn{
+			stdoutstderr: stdoutstderr,
+			err:          err,
+		}:
+		}
+	}(ch)
+
+	if executionTimeout.Seconds() > 0 {
+		ctx, cancelFunc := context.WithTimeout(context.Background(), executionTimeout)
+		defer cancelFunc()
+		go func() {
+			select {
+			case <-funcContext.Done():
+			case <-ctx.Done():
```

Comment:
You can probably avoid the double context by simply closing the session after the timeout, instead of having the session closed eventually with the `defer`.